oxidation/libparsec/crates/platform_local_db/src/test_utils.rs

```diff
@@ -1,63 +1,128 @@
+                                path
+                            )
+                            // In theory here we should reset the entry state (given we set
+                            // it before doing the check), however in the current case both
+                            // state have the same `Opened` value so all is already good ;-)
+                        }
+                        ClosableInMemoryDB::Closed(db) => {
+                            // The database has been closed by the one that created it.
+                            // Of course in-memory db loose all data on close so it has
+                            // been kept on the side instead, and now is time to re-use it !
+                            Some(db)
+                        }
+                    }
+                }
+
+                // This database is brand new
+                Entry::Vacant(entry) => {
+                    // Now create the database and register it path as opened
+                    let conn = SqliteConnection::establish(":memory:")
+                        .expect("Cannot create in-memory database");
+                    entry.insert(ClosableInMemoryDB::Opened);
+                    Some(SqliteExecutor::spawn(conn))
+                }
+            }
         }
-    };
 
-    Some(res.map(|executor| LocalDatabase {
-        executor: Arc::new(RwLock::new(Some(executor))),
-    }))
+        TestDBStrategy::OnDisk => None,
+    }
 }
 
-/// Enable or not to have [LocalDatabase] in memory.
-///
-/// Note: Already openned connection would remain unchanged.
-pub fn toggle_local_db_in_memory(enabled: bool) {
-    OPEN_CONNECTION_IN_MEMORY.store(enabled, Ordering::SeqCst)
+pub(crate) fn return_sqlite_in_memory_db(path: &str, db: SqliteExecutor) {
+    let mut strategy = TEST_DB_STRATEGY.lock().expect("Mutex is poisoned");
+    if let TestDBStrategy::InMemory(dbs) = &mut *strategy {
+        if let Entry::Occupied(mut entry) = dbs.entry(path.to_string()) {
+            if let ClosableInMemoryDB::Opened = entry.get() {
+                entry.insert(ClosableInMemoryDB::Closed(db));
+                return;
+            }
+        }
+    }
+    panic!(
+        "Inconsistent state with SQLite in-memory mock:\npath: {}\ndbs: {:?}",
+        path, *strategy
+    )
 }
 
-/// Will clear saved database opened in memory.
+/// Enable or not to open databases in memory.
 ///
-/// Note: If [LocalDatabase] are still running, those connection would remain openned until it's dropped.
-pub fn clear_local_db_in_memory() {
-    LOCAL_DB_IN_MEMORY
-        .lock()
-        .expect("Mutex is poisoned")
-        .clear();
+/// Note: Already opened connections remain unchanged.
+pub fn test_toggle_local_db_in_memory_mock(enabled: bool) {
+    let mut strategy = TEST_DB_STRATEGY.lock().expect("Mutex is poisoned");
+    match (enabled, &mut *strategy) {
+        (false, s @ TestDBStrategy::InMemory(_)) => {
```

Comment:
You don't need `s @` if you don't match a function call

Reply:
I need to first match that strategy contains a InMemory, then change the value of strategy, you mean there is a way to do it without using `s @`

Reply:
it's `strategy` itself